dlib/pixel.h

```diff
@@ -1223,6 +1223,8 @@ namespace dlib
             c1.b = src.blue/255.0;
             c2 = RGB2HSL(c1);
 
+            // Bug: undefined behaviour trigged. This is what ubsan says:
+            // runtime error: -14.6786 is outside the range of representable values of type 'unsigned char'
```

Comment:
Did this actually happen?  What input causes this?

Reply:
Yes, try running the newly added unit test with the inputs in the problematic directory, and you will see it. (compiling with clang and undefined behaviour sanitizer)

Reply:
Can you find out the exact pixel value that triggers it?  That would make fixing and testing it easier for me :)

Reply:
This will trigger the undefined behaviour, but it won't be detected by DLIB_TEST, because the UB is inside the function. I put this inside the unit test in pixel.cpp
```cpp
            rgb_pixel src;
            src.red=246;
            src.green=170;
            src.blue=245;
            hsi_pixel dest;
            assign_pixel(dest,src);
            DLIB_TEST(dest.h>=0);
            DLIB_TEST(dest.h<=360);
```

Reply:
I think the missing part is the check for hue<0, it should be wrapped around by adding 360. Looking at the code, it seems like it gives -60<hue<300

Looking at the wikipedia entry, it seems like the function could be simplified a bit:
https://en.wikipedia.org/wiki/HSL_and_HSV#From_RGB

Reply:
I pushed a minimal fix.